### PR TITLE
EI-528 - Revert to go-ethereum for secp256k functionality (introduced…

### PR DESCRIPTION
**Pull request by vtermanis**
… in 398057a)

- github.com/fomichev/secp256k1 has negative performance impact
- vulnerability associated with go-ethereum is not relevant (i.e. code not touched)
  See also https://github.com/ethereum/go-ethereum/issues/23866
- Bump github.com/ethereum/go-ethereum to v1.10.14

**Notes**:
- ~v1.16 compatible via `go mod tidy -compat=1.17 -compat=1.16`~ Now v1.17 targeted, with `go mod tidy -compat=1.17`
- Not a straight-up revert since keeping minor refactoring which also was in 398057a

### CODE REVIEW

**Comment by vtermanis** [comment]
I suggest we add open an issue with the following text to explain why the related CVE is a non-issue:

---

[CVE-2021-43668](https://nvd.nist.gov/vuln/detail/CVE-2021-43668) ([github.com/ethereum/go-ethereum](https://github.com/ethereum/go-ethereum))

`iotics-identity-go library` uses `go-ethereum` but only a very small set of library functions from it, namely:
- `go-ethereum/crypto.UnmarshalPubkey()` (code)
- `go-ethereum/crypto/secp256k1.BitCurve`  (tests only)

The aforementioned CVE affects only `go-ethereum`'s `geth` executable and its use of [github.com/syndtr/goleveldb](https://github.com/syndtr/goleveldb) . Because of this, `iotics-identity-go` is not affected by this security issue:

```shell
$ go mod why -m github.com/ethereum/go-ethereum
# github.com/ethereum/go-ethereum
github.com/Iotic-Labs/iotics-identity-go/pkg/crypto
github.com/ethereum/go-ethereum/crypto

$ go mod why -m github.com/syndtr/goleveldb
# github.com/syndtr/goleveldb
(main module does not need module github.com/syndtr/goleveldb)
```

### Ticket references
- https://github.com/ethereum/go-ethereum/issues/23866
- https://github.com/syndtr/goleveldb/issues/373

File: pkg/crypto/keys.go [hunk 1]
```diff
@@ -49,7 +28,7 @@ func GetPublicKeysFromPrivateKey(privateKey *ecdsa.PrivateKey) ([]byte, string,
 		return nil, "", errors.New("error casting public key to ECDSA")
 	}
 
-	publicKeyDer := elliptic.Marshal(secp256k1.SECP256K1(), publicKeyECDSA.X, publicKeyECDSA.Y)
+	publicKeyDer := ethcrypto.FromECDSAPub(publicKeyECDSA)
```
Comment (smartrics) [inline pkg/crypto/keys.go:31]: this method is simply
`func FromECDSAPub(pub *ecdsa.PublicKey) []byte {
	if pub == nil || pub.X == nil || pub.Y == nil {
		return nil
	}
	return elliptic.Marshal(S256(), pub.X, pub.Y)
}`

maybe we can inline it in our SDK being just a validation on the elliptic marshaller?

File: pkg/crypto/keys.go [hunk 2]
```diff
@@ -58,11 +37,5 @@ func GetPublicKeysFromPrivateKey(privateKey *ecdsa.PrivateKey) ([]byte, string,
...
 func GetPublicKeyFromBase58(publicBase58 string) (*ecdsa.PublicKey, error) {
 	publicKeyBytes := base58.DecodeAlphabet(publicBase58, base58.BTCAlphabet)
-
-	curve := secp256k1.SECP256K1()
-	x, y := elliptic.Unmarshal(curve, publicKeyBytes)
-	if x == nil {
-		return nil, fmt.Errorf("invalid secp256k1 public key")
-	}
-	return &ecdsa.PublicKey{Curve: curve, X: x, Y: y}, nil
+	return ethcrypto.UnmarshalPubkey(publicKeyBytes)
```
Comment (smartrics) [inline pkg/crypto/keys.go:40]: [this](https://github.com/ethereum/go-ethereum/blob/127ce93db4f5af4d7ada3066a15942a0ab59a948/crypto/crypto.go#L168) is a utility method that's only wrapping someother code

Comment (smartrics) [reply pkg/crypto/keys.go:40]: I wonder if there's scope to inline it (copy/paste)

**Review by smartrics — COMMENTED** [review]
@vtermanis if we were to copy paste the three methods we would not really dup any code since the methods are just utilities (there's no logic at all except validation)
we should then avoid go-ethereum altogether and make it simpler overall
this is preached on the basis that we're only reusing utilities rather than logic

File: pkg/crypto/keys.go (hunk 1 above)
Comment (vtermanis) [reply pkg/crypto/keys.go:31]: But: `S256()` is something which belongs to `go-ethereum` - what would the benefit be of in-lining this?

**Comment by vtermanis** [comment]
> we should then avoid go-ethereum altogether and make it simpler overall

I don't think we do avoid it because of the usage of of `S256` from [here](https://github.com/ethereum/go-ethereum/blob/127ce93db4f5af4d7ada3066a15942a0ab59a948/crypto/secp256k1/curve.go#L296)

File: pkg/crypto/keys.go (hunk 2 above)
Comment (vtermanis) [reply pkg/crypto/keys.go:40]: (As above - uses `S256` which is a `go-ethereum/crypto` definition.) The point is that the general golang `crypto` methods don't implement `secp256k1`)

File: pkg/crypto/keys.go (hunk 1 above)
Comment (smartrics) [reply pkg/crypto/keys.go:31]: gotcha - i missed it

**Review by smartrics — APPROVED** [review]

**Comment by smartrics** [comment]
Text in [this](https://github.com/Iotic-Labs/iotics-identity-go/pull/33#issuecomment-1007321625) comment looks good to me

**Pull request merged**